src/lib/plugins/login.js

```diff
@@ -36,46 +31,31 @@ module.exports.server = function (serv, options) {
   })
 }
 
-module.exports.player = function (player, serv, settings) {
+module.exports.player = async function (player, serv, settings) {
   const Item = require('prismarine-item')(settings.version)
   const mcData = require('minecraft-data')(settings.version)
 
-  let playerSavedInventoryItems = []
-  let playerSavedPosition = []
-  let playerSavedRotation = []
-  let playerSavedOnGround = true
+  let playerData
 
   async function addPlayer () {
     player.type = 'player'
     player.crouching = false // Needs added in prismarine-entity later
     player.op = settings['everybody-op'] // REMOVE THIS WHEN OUT OF TESTING
     player.username = player._client.username
     player.uuid = player._client.uuid
-    try {
-      const playerDataFile = await fsReadFile(`${settings.worldFolder}/playerdata/${player.uuid}.dat`)
-      const playerData = (await nbtParse(playerDataFile)).value
-      player.health = playerData.Health.value
-      player.food = playerData.foodLevel.value
-      player.gameMode = playerData.playerGameType.value
-      player.xp = playerData.XpTotal.value
-      player.heldItemSlot = playerData.SelectedItemSlot.value
-      playerSavedPosition = playerData.Pos.value.value
-      playerSavedOnGround = Boolean(playerData.OnGround.value)
-      playerSavedRotation = playerData.Rotation.value.value
-      playerSavedInventoryItems = playerData.Inventory.value.value
-    } catch (err) {
-      // No player data file / other error
-      player.health = 20
-      player.food = 20
-      player.heldItemSlot = 0
-    }
+
+    await player.findSpawnPoint()
+
+    playerData = await playerDat.read(player.uuid, player.spawnPoint, settings.worldFolder)
+    Object.keys(playerData.player).forEach(k => { player[k] = playerData.player[k] })
+
     serv.players.push(player)
     serv.uuidToPlayer[player.uuid] = player
     player.loadedChunks = {}
   }
 
   function updateInventory () {
-    playerSavedInventoryItems.forEach((item) => {
+    playerData.inventory.forEach((item) => {
```

Comment:
why not do this in addPlayer ? (that way you can make playerData local for addPlayer function)

Reply:
It seems like player.inventory.updateSlot doesn't work at the time when addPlayer is called

Reply:
player.inventory.updateSlot works only after sendLogin

Reply:
ok